CromwellRefdiskManifestCreator/src/main/java/org/broadinstitute/manifestcreator/CromwellRefdiskManifestCreatorApp.java

```diff
@@ -0,0 +1,151 @@
+import java.util.concurrent.Executors;
+import java.util.stream.Stream;
+import java.util.zip.CRC32;
+
+public class CromwellRefdiskManifestCreatorApp {
+
+  private static final Logger logger = LogManager.getLogger(CromwellRefdiskManifestCreatorApp.class);
+
+  public static void main(String[] args) throws IOException, InterruptedException {
+    Configurator.setRootLevel(Level.DEBUG);
+
+    if (args.length < 4) {
+      logger.error("Wrong number of parameters.");
+      logger.error("Usage: java -jar cromwell-refdisk-manifest-creator-app.jar <number of parallel threads> " +
+              "<image_name> <directory_path_to_scan> <output file path");
+      System.exit(1);
+    }
+
+    String directoryToScan = args[2];
+    File rootDirectory = new File(directoryToScan);
+    if (!rootDirectory.exists() || !rootDirectory.isDirectory()) {
+      logger.error("Root directory " + directoryToScan + " doesn't exist.");
+      System.exit(1);
+    }
+
+    int totalNumberOfFiles = countFiles(rootDirectory);
+    logger.info("Parsing " + totalNumberOfFiles + " files");
+
+    ReferenceDiskManifest manifest = new ReferenceDiskManifest();
+    String imageName = args[1];
+    manifest.setImageIdentifier(imageName);
+
+    int nThreads = Integer.parseInt(args[0]);
+    CountDownLatch countDownLatch = new CountDownLatch(nThreads);
+    ExecutorService executorService = Executors.newFixedThreadPool(nThreads);
+
+    Path rootPath = rootDirectory.toPath();
+    Stack<File> stack = new Stack<>();
+    stack.push(rootDirectory);
+    List<ReferenceFile> referenceFiles = Collections.synchronizedList(new ArrayList<>());
+    for (int i = 0; i < nThreads; i++) {
+      executorService.submit(() -> {
+        try {
+          doWork(rootPath, referenceFiles, stack, totalNumberOfFiles, countDownLatch);
+        } catch (CRC32Exception e) {
+          logger.error("Error occurred. Shutting down.", e);
+          executorService.shutdownNow();
+          System.exit(1);
+        }
+      });
+    }
+
+    countDownLatch.await();
+
+    String manifestFilePath = args[3];
+    File manifestFile = new File(manifestFilePath);
+    if (manifestFile.exists() || !manifestFile.createNewFile()) {
+      logger.error("File " + manifestFile.getAbsolutePath() + " already exists or cannot be created.");
+    } else {
+      referenceFiles.forEach(referenceFile -> manifest.getFiles().add(referenceFile));
+      new ObjectMapper().writerWithDefaultPrettyPrinter().writeValue(manifestFile, manifest);
+      logger.info("Finished. Total number of reference files: " + manifest.getFiles().size());
+    }
+
+    executorService.shutdown();
+  }
+
+  private static void doWork(Path rootPath, List<ReferenceFile> accumulator, Stack<File> stack, int totalNumberOfFiles, CountDownLatch countDownLatch) throws CRC32Exception {
+    while(!stack.isEmpty() || accumulator.size() != totalNumberOfFiles) {
+      if (stack.isEmpty()) {
+        continue;
+      }
+
+      File curFile = stack.pop();
```

Comment:
is there a race here if another thread has popped the stack to emptiness?

Reply:
Yes, thanks